### PR TITLE
Apex with DataParallel fixes

### PR DESCRIPTION
**Pull request by and-kul**
## Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [ ] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [ ] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [ ] Did you check the docs with `make check-docs`?
- [ ] Did you write any new necessary tests?
- [ ] Did you check that your code passes the unit tests `pytest .` ? 
- [ ] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?

## Description

1. Allow apex for key-value model
2. Allow apex+opt_level="O2"+DataParallel

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: catalyst/utils/components.py [hunk 1]
```diff
@@ -19,14 +26,69 @@
...
+    return model
+
+
+# apex issue https://github.com/deepset-ai/FARM/issues/210
+# solution: https://github.com/NVIDIA/apex/issues/503#issuecomment-566181771
+def _wrap_into_data_parallel_with_apex(
+    model: RunnerModel, optimizer: Optimizer, distributed_params: Dict
+):
+    if isinstance(model, nn.Module):
+        model = nn.Sequential(model)
```
Comment (Scitator) [inline catalyst/utils/components.py:80]: why are we checking the model for ``nn.Module``, but cast it to ``nn.Sequential`` after that?

File: catalyst/utils/components.py [hunk 2]
```diff
@@ -19,14 +26,69 @@
...
+        model, optimizer = initialize_apex(
+            model, optimizer, **distributed_params
+        )
+        model = torch.nn.DataParallel(model[0])
+        model = _patch_forward(model)
```
Comment (Scitator) [inline catalyst/utils/components.py:85]: okay, that looks like really tricky solution

File: catalyst/utils/components.py [hunk 3]
```diff
@@ -19,14 +26,69 @@
...
+        lambda tensor: tensor.to(
+            apex.amp._amp_state.opt_properties.options.get(
+                "cast_model_outputs", torch.float32
+            )
+        )
+        if tensor.is_floating_point()
+        else tensor
+    )
+
+    model.forward = lambda *args, old_fwd=model.forward, input_caster=input_caster_lambda, output_caster=output_caster_lambda, **kwargs: apex.amp._initialize.applier(
```
Comment (Scitator) [inline catalyst/utils/components.py]: is it possible to use ``**kwargs`` with this hack?

Comment (and-kul) [reply catalyst/utils/components.py]: yep

**Comment by and-kul** [comment]
Changelog is updated, codestyle is fixed

**Pull request merged**